crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,171 @@
+///
+/// An item stack includes an item type, an amount and a bunch o properties (enchantments, etc.)
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ItemStack {
+
+    // TODO Migrate all the codegen boilerplate.
+    /// The item type of this `ItemStack`.
+    pub item: Item,
+
+    /// The number of items in the `ItemStack`.
+    pub count: u32,
+
+    /// Damage to the item, if it's damageable.
+    pub damage: Option<u32>,
+
+    /// Repair cost of the item, if it's repairable.
+    pub repair_cost: Option<u32>,
+
+    /// Enchantments applied to the item.
+    pub enchantments: Vec<Enchantment>,
+
+    // TODO Add other properties like title and lore.
+}
+
+impl ItemStack {
+
+    /// Creates a new `ItemStack`
+    pub fn new(item: Item, count: u32) -> Self {
+        Self {
+            item,
+            count,
+            damage: item.durability().map(|_| 0)
+        }
+    }
+
+    /// Returns whether the given item stack has
+    /// the same type as (but not necessarily the same
+    /// amount as) `self`.
+    pub fn has_same_type(&self, other: &Self) -> bool {
+        self.item == other.item
+    }
+
+    /// Returns whether the given item stack has the same damage
+    /// as `self`.
+    pub fn has_same_damage(&self, other: &Self) -> bool {
+        self.damage == other.damage
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same count as (but not necessarily the same
+    /// type as) `self`.
+    pub fn has_same_count(&self, other: &Self) -> bool {
+        self.count == self.count
+    }
+
+    /// Returns whether the given `ItemStack` has the same
+    /// type and count as `self`.
+    pub fn has_same_type_and_count(&self, other: &Self) -> bool {
+        self.item == other.item && self.count == other.count
+    }
+
+    /// Returns whether the given `ItemStack` has
+    /// the same type and damage as `self`.
+    pub fn has_same_type_and_damage(&self, other: &Self) -> bool {
+        self.item == other.item && self.damage == other.damage
+    }
+
+    /// Returns the item type for this `ItemStack`.
+    pub fn item(&self) -> Item {
+        self.item
+    }
+
+    /// Returns the number of items in this `ItemStack`.
+    pub fn count(&self) -> u32 {
+        self.count
+    }
+
+    /// Adds more items to this `ItemStack`. Returns the new count.
+    pub fn add(&mut self, count: u32) -> u32 {
+        self.count += count;
+        self.cout
+    }
+
+    /// Removes some items from this `ItemStack`.
+    pub fn remove(&mut self, count: u32) -> Result<u32, ()> {
+        self.count = match self.count.checked_sub(count) {
+            Some(count) => count,
+            None => return Err(())
+        };
+        Ok(self.count)
+    }
+
+    /// Sets the item type for this `ItemStack`.
+    pub fn set_item(&mut self, item: Item) {
+        self.item = item
+    }
+
+    /// Sets the count for this `ItemStack`.
+    pub fn set_count(&mut self, count: u32) {
+        self.count = count

```

Comment:
Stack limit should be respected.

Reply:
I think this method should not respect the item stack count. I don't see when this method will be used by the server natively (the server will use `add` and `remove`) methods. By adding this `set_count` method you allow future plugins to generate stacks of items over their specified stack size. This is doable and already implemented on some servers, for example in CubeCraft it's easy to find loot made of stacks of 20 or 25 eggs.

![imagen](https://user-images.githubusercontent.com/36012050/109058289-1eb89880-76e3-11eb-97b0-5630e3f2190a.png)

I think we should not close completely the ability to generate stacks of "illegal" stack count. I would leave this method unchecked and maybe update the docs to clarify this while restricting `add` and `remove`.

Reply:
Maybe add an unsafe_set and unsafe_add?

Reply:
That will do too. Nice idea 😄

Reply:
Implemented `unchecked_add`, `unchecked_set_count` and `unchecked_set_item`. Restrained `set_count` and `set_item`.